docs/testplan/PFC_Congestion_Oversubscription_Test_Plan.md

```diff
@@ -0,0 +1,318 @@
+
+This experiment needs the following eight steps:
+1. Start the test data traffic from IXIA Tx ports 1; the combined line rate will be 60% of the egress port capacity.
+2. Start the background traffic from IXIA Tx ports 2; the additional increase in bandwidth will be 60%, for a total oversubscription of 120%.
+3. After a few seconds, measure the received bandwidth per flow on the IXIA Rx port 3: test data traffic and background traffic with 20% of bandwidth should not lose any packet. Internally the router should be dropping lossy frames from the streams causing the over-subscription (Port 2). There should be PFC pause messages being received at the IXIA transmit port 1 stream 2 causing the over-subscription.
+4.	Stop all traffic from the two IXIA ports 1 & 2
+5.	On the IXIA Rx port ensure that:
+a.	The received lossless traffic streams on IXIA port 3 show no missing and no out-of-order frames.
+b.	The lossy traffic streams causing the oversubscription had some frame drops.
+6.	If possible, query the DUT drop counters to verify that the lossy stream drops were accounted for.
+7.	Repeat the test by adding a impairment (Network emulator) device between Tx port and DUT ingress port to simulate various cable lengths and validate that PFC headroom is sufficient for variable cable lengths.
+8.	Verify that the test results are same for varying cable lengths.
+
+
+
+### Test case # 4 – Many-to-One with Fluctuating Lossless Traffic Congestion
+#### Test objective
+Test the dynamic response of lossy and lossless streams destined to a single egress port when the lossless streams cause periods of over-subscription.
+PFC pause frames should automatically be generated by the device under test (DUT) to throttle back the lossless traffic sources causing the over-subscription, avoiding any lossless traffic drops. The PFC pause frame generation should stop once the over-subscription condition subsides.
+
+#### Background
+This test uses a similar diagram and description of the multi-ASIC router as the ‘Many-to-One Lossless Traffic Congestion’ test case (it is indeed a superset of that test).
+In this variant, we will use all the lossy and lossless traffic priorities simultaneously, and will expect an even distribution of traffic across all eight priorities even under periods of oversubscription caused by lossless traffic streams. The assumption is that the SONiC default scheduler will allocate same weights to all eight priorities in its WRR scheduler with no strict priority (SP) priorities configured.
+
+#### Testbed Setup
+
+The test bed setup consists of three IXIA ports and a SONiC router as the device under test (DUT). All IXIA ports should have the same line speed.  The three DUT ports will be chosen at random across all the available ports in the router in order to eventually cover combinations of ports within the same line card and across line cards (or ports within the same ASIC vs. different ASICs on a single-card, multi-ASIC design).

```

Comment:
Can we use mix port speeds here ?

Reply:
We can use different port speeds. But to validate the expected outcome, we are using same speed modes. If not it will be very difficult to meet the expectations for different speed modes.